Fix inventory list

jniles opened the pull request:
This PR refactors and fixes the inventory module with most suggestions from #926 and #920.  Specifically, it
 1. Creates `inventory.routes.js` and transitions the inventory to state-based routing.
 2. Cleans up the inventory create/update modal greatly - up to 2.x standards.
 3. Cleans up the inventory service with PrototypeApiService.

Closes #926.  Closes #920.


-----
Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

Commit:
fix(exchange): broke home page link

This commit fixes the broken update exchange rate link on the home page.
It now correctly points to the enterprise management page.

Commit:
fix(inventory): eliminate pagination

This commit removes pagination in the inventory list.

Commit:
fix(inventory): create/update error states

This commit ensures that the create/update modal properly handles error
state and reporting.  It also improves the UI slightly.

Commit:
fix(inventory): order inventory items on download

This commit changes the server to send back an ordered list of inventory
items, ordered by their code.

Commit:
refactor(inventory): ui-route for create state

This commit breaks out the inventory routes into their own
inventory.routes.js file.  This file sets up the create state in a modal
for the inventory items to make it easier to navigate to.  Eventually,
it should implement a checkbox to add many items, such as the account
page.

It also improves the create form by giving it a bh-loading-button, and
cleaning up some of the validation.  Notify.handleError has now been
implemented, instead of the previous bug in error handling.

Commit:
refactor(inventory): cosmetic improvements

This commit adds a headercrumb to the inventory modal, as well as
refactors the create/edit state handling within the modal controller.
Finally, it ensures that a notification is shown when the create/update
state successfully completes.

Commit:
fix(report): print-friendly tables

This commit adds rules to enable print-friendly tables in wkhtmltopdf.
It adds page break rules to indicate that pages should not break mid-row
in an overflow.

Commit:
refactor(inventory): refactor modal to use $state

This commit fixes some important bugs in the inventory update modal.  In
particular, the modal will wait to bind the inventory uuid in the UPDATE
state until it has all the type, unit, and group information.

It also completes the $state-based routing for the inventory modal.

Commit:
refactor(inventory): use ProtypeApiService

This commit migrates a lot of the dependent Inventory Services to use
PrototypeApiService.

sfount commented on client/src/partials/inventory/configuration/units/modals/actions.modal.js:
```diff
@@ -34,19 +32,13 @@ function InventoryUnitActionsModalController(InventoryUnit, Notify, Instance, Da
-      return res;
-    })
-    .catch(Notify.errorHandler);
+      .catch(Notify.errorHandler);
```
I don't think that `Notify.errorHandler` exists. I know this is a re-factor but it would be good to clean these up now - this error will be thrown as a client error silently. 

🐛

sfount commented on client/src/partials/inventory/configuration/units/modals/actions.modal.js:
```diff
@@ -34,19 +32,13 @@ function InventoryUnitActionsModalController(InventoryUnit, Notify, Instance, Da
-      return res;
-    })
-    .catch(Notify.errorHandler);
+      .catch(Notify.errorHandler);
```
:bug:

jniles replied:
Ha!

sfount commented on client/src/partials/inventory/configuration/units/units.js:
```diff
@@ -25,35 +25,34 @@ function InventoryUnitsController(InventoryUnit, Notify, Modal) {
+        }
+      })
+      .then(startup)
+      .catch(Notify.errorHandler);
```
:bug:

sfount commented on client/src/partials/inventory/configuration/units/units.js:
```diff
@@ -25,35 +25,34 @@ function InventoryUnitsController(InventoryUnit, Notify, Modal) {
-    .then(startup)
-    .catch(Notify.errorHandler);
+      .then(function (res) {
+        if (res.id) {
```
Do you know what ` if (res.id)` does for you here? What if there is no `id` on `res`?

jniles replied:
Yes, the modal will return an id if a record was created or updated.  I'm not sure if a case were it wouldn't return an `id`, except by a broken API.

Do you have a suggestion to make this clearer?

sfount commented on client/src/partials/inventory/configuration/units/units.js:
```diff
@@ -25,35 +25,34 @@ function InventoryUnitsController(InventoryUnit, Notify, Modal) {
+      .then(function (list) {
+        vm.unitList = list;
+      })
+      .catch(Notify.errorHandler);
```
🐛

sfount reviewed (commented):
The changes here LGTM. Some really great usability changes that make the UX much better. 

I can confirm that it should close the issues referenced. 

I have made a number of comments however none of them are blocking.

Commit:
fix(Notify): errorHandler --> handleError

This commit changes all `Notify.errorHandler()` calls to
`Notify.handleError()`, throughout the entire application.
